ui/app/components/auth-jwt.js

```diff
@@ -16,129 +17,149 @@ const ERROR_MISSING_PARAMS =
+    this._authType = this.args.selectedAuthType;
+  }
+
+  get getWindow() {
+    return window;
+  }
+
+  fetchRole = restartableTask(async (wait) => {
+    // if we have a custom path is inputted use that,
+    // otherwise fallback to type (which is the default path)
+    const path = this.args.selectedAuthPath || this.args.selectedAuthType;
+
+    // task is `restartable` so if the user starts typing again,
+    // it will cancel and restart from the beginning.
+    if (wait) await timeout(wait);
 
-  async fetchRole() {
-    const path = this.selectedAuthPath || this.selectedAuthType;
     const id = JSON.stringify([path, this.roleName]);
-    this.setProperties({ role: null, errorMessage: null, isOIDC: true });
+
+    this.fetchedRole = null;
+    this.errorMessage = null;
+    this.isOIDC = true;
 
     try {
-      const role = await this.store.findRecord('role-jwt', id, {
-        adapterOptions: { namespace: this.namespace },
+      this.fetchedRole = await this.store.findRecord('role-jwt', id, {
+        adapterOptions: { namespace: this.args.namespace },
       });
-      this.set('role', role);
     } catch (e) {
       const error = (e.errors || [])[0];
       const errorMessage =
         e.httpStatus === 400 ? 'Invalid role. Please try again.' : `Error fetching role: ${error}`;
       // assume OIDC until it's known that the mount is configured for JWT authentication via static keys, JWKS, or OIDC discovery.
-      this.setProperties({ isOIDC: error !== ERROR_JWT_LOGIN, errorMessage });
+      // if the mount is configured for JWT this specific error is returned.
+      this.isOIDC = error !== ERROR_JWT_LOGIN;
+      this.errorMessage = errorMessage;
     }
-  },
+  });
 
   cancelLogin(oidcWindow, errorMessage) {
     this.closeWindow(oidcWindow);
     this.handleOIDCError(errorMessage);
-  },
+  }
 
   closeWindow(oidcWindow) {
     this.watchPopup.cancelAll();
     this.watchCurrent.cancelAll();
     oidcWindow.close();
-  },
+  }
 
   handleOIDCError(err) {
-    this.onLoading(false);
     this.prepareForOIDC.cancelAll();
-    this.onError(err);
-  },
+    this.args.onError(err);
+  }
 
   // NOTE TO DEVS: Be careful when updating the OIDC flow and ensure the updates
   // work with implicit flow. See issue https://github.com/hashicorp/vault-plugin-auth-jwt/pull/192
-  prepareForOIDC: task(function* (oidcWindow) {
-    const thisWindow = this.getWindow();
-    // show the loading animation in the parent
-    this.onLoading(true);
+  prepareForOIDC = task(async (oidcWindow) => {
+    const thisWindow = this.getWindow;
+
     // start watching the popup window and the current one
     this.watchPopup.perform(oidcWindow);
     this.watchCurrent.perform(oidcWindow);
     // wait for message posted from oidc callback
     // see issue https://github.com/hashicorp/vault/issues/12436
     // ensure that postMessage event is from expected source
+    // eslint-disable-next-line no-constant-condition
```

Comment:
I attempted refactoring this to avoid `while (true)` but it introduced some regressions (caught by the updated acceptance version of [this](https://github.com/hashicorp/vault/pull/30130/files#diff-3a8115340d3e16976ef6c8f1d710ccdc93efb6726902a748366e258267a436deL55) test, thanks @zofskeez for the excellent test coverage 😎 )

This syntax is actually suggested by the [ember-concurrency docs](https://ember-concurrency.com/docs/events/). So I think it make sense to leave as is. I can explore refactoring with the new auth flow.
<img width="917" alt="Screenshot 2025-04-02 at 1 30 55 PM" src="https://github.com/user-attachments/assets/96b4473f-5110-49c4-a13c-61dd2eeb7c31" />

Reply:
Yay for tests 🎉! I think it's fine since that's how the task was originally set up and like you mentioned the docs suggest that pattern.